cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -2851,3 +2851,20 @@
 (rule (lower (has_type (and (fits_in_64 ty) (ty_int _))
                   (atomic_cas flags address expected replacement)))
       (x64_cmpxchg ty expected replacement (to_amode flags address (zero_offset))))
+
+;; Rules for `atomic_rmw` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; This is a simple, general-case atomic update, based on a loop involving
+;; `cmpxchg`.  Note that we could do much better than this in the case where the
+;; old value at the location (that is to say, the SSA `Value` computed by this
+;; CLIF instruction) is not required.  In that case, we could instead implement
+;; this using a single `lock`-prefixed x64 read-modify-write instruction.  Also,
+;; even in the case where the old value is required, for the `add` and `sub`
+;; cases, we can use the single instruction `lock xadd`.  However, those
+;; improvements have been left for another day. TODO: filed as
+;; https://github.com/bytecodealliance/wasmtime/issues/2153.
+
+(rule (lower (has_type (and (fits_in_64 ty) (ty_int _))
+                  (atomic_rmw flags op address input)))
+      ;; TODO: `flags` is unused--should it be?
```

Comment:
Can we assert something about `flags`? (We could pass it into `x64_atomic_rmw_seq` and match only the flags that we expect, or something like that...)

Reply:
Not sure what we should assert here.

Reply:
Ok, take a look at 3799054, which calculates the amode at this point instead of in `emit.rs`. I believe this is the more conventional way to do things in ISLE now, but unfortunately I see failures related `reg.is_real()` so the `mem.finalize...` part must not be right.

Reply:
Ah, I think this is missing the corresponding use of the amode in the `get_operands` method; once that's added (in the same order as the `allocs` invocations) it should probably work?

Reply:
Hm, that fixes one problem but now I see `wasm trap: out of bounds memory access` from `threads.wast`. `mem.get_operands()` ends up doing a `collector.reg_use()` instead of `collector.reg_late_use()` like we were doing previously--seems like this could be the problem?

Reply:
Ok, that seems to work.